Integrations/MicrosoftGraphListener/MicrosoftGraphListener.py

```diff
@@ -0,0 +1,986 @@
+        self._app_name = app_name
+        self._mailbox_to_fetch = mailbox_to_fetch
+        self._folder_to_fetch = folder_to_fetch
+        self._first_fetch_interval = first_fetch_interval
+        self._emails_fetch_limit = emails_fetch_limit
+
+    def _get_access_token(self):
+        """
+        Obtains access and refresh token from Oprxoy server. Access token is used and stored in the integration context
+        until expiration time. After expiration, new refresh token and access token are obtained and stored in the
+        integration context.
+
+        :return: Access token that will be added to authorization header
+        :rtype: ``str``
+        """
+        integration_context = demisto.getIntegrationContext()
+        access_token = integration_context.get('access_token')
+        valid_until = integration_context.get('valid_until')
+        if access_token and valid_until:
+            if epoch_seconds() < valid_until:
+                return access_token
+
+        oproxy_response = requests.post(
+            self._token_retrieval_url,
+            json={
+                'app_name': self._app_name,
+                'registration_id': self._auth_id,
+                'encrypted_token': get_encrypted(self._refresh_token, self._enc_key)
+            },
+            verify=self._verify
+        )
+
+        if oproxy_response.status_code not in {200, 201}:
+            msg = 'Error in authentication. Try checking the credentials you entered.'
+            try:
+                demisto.info('Authentication failure from server: {} {} {}'.format(
+                    oproxy_response.status_code, oproxy_response.reason, oproxy_response.text))
+                err_response = oproxy_response.json()
+                server_msg = err_response.get('message')
+                if not server_msg:
+                    title = err_response.get('title')
+                    detail = err_response.get('detail')
+                    if title:
+                        server_msg = f'{title}. {detail}'
+                if server_msg:
+                    msg += ' Server message: {}'.format(server_msg)
+            except Exception as ex:
+                demisto.error('Failed parsing error response - Exception: {}'.format(ex))
+            raise Exception(msg)
+        try:
+            gcloud_function_exec_id = oproxy_response.headers.get('Function-Execution-Id')
+            demisto.info(f'Google Cloud Function Execution ID: {gcloud_function_exec_id}')
+            parsed_response = oproxy_response.json()
+        except ValueError:
+            raise Exception(
+                'There was a problem in retrieving an updated access token.\n'
+                'The response from the Oproxy server did not contain the expected content.'
+            )
+        access_token = parsed_response.get('access_token')
+        expires_in = parsed_response.get('expires_in', 3595)
+        current_refresh_token = parsed_response.get('refresh_token')
```

Comment:
This may be empty. oproxy may wish not to return a new refresh token if there wasn't a change. Please take this into account.

Reply:
As I mentioned in the issue, that isn't possible scenario.